Hola Eduardo!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
</div>

In [3]:
# Importarción de librerias necesarias para el analisis basico de ML
import pandas as pd
from sklearn.model_selection import train_test_split

In [4]:
# cargar archivo con el que estaremos trabajando
df = pd.read_csv('C:/Users/jedua/OneDrive/Documentos/TTProyectos/users_behavior.csv')

# Separacion de características y objetivo antes del entrenamiento del modelo 
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

#Muestra de la informacion general.

print("este df es una muestra general del DF principal",df.head(),"\n")

print("este df son las caracteristicas",features.head(),"\n")
print("este es el objetivo",target.head())


este df es una muestra general del DF principal    calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0 

este df son las caracteristicas    calls  minutes  messages   mb_used
0   40.0   311.90      83.0  19915.42
1   85.0   516.75      56.0  22696.96
2   77.0   467.66      86.0  21060.45
3  106.0   745.53      81.0   8437.39
4   66.0   418.74       1.0  14502.75 

este es el objetivo 0    0
1    0
2    0
3    1
4    0
Name: is_ultra, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con la carga de datos y separando con la variable objetivo a predecir
</div>

In [5]:
# Desglose de DF en training, validation, and test sets (70% training, 15% validation, 15% test)
features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size=0.30, random_state=54321)
features_valid, features_test, target_valid, target_test = train_test_split(features_temp, target_temp, test_size=0.50, random_state=54321)

# Mostrar el tamaño de cada SET de datos
print("Tamaño del conjunto de entrenamiento (features_train):", features_train.shape)
print("Tamaño del conjunto de validación (features_valid):", features_valid.shape)
print("Tamaño del conjunto de prueba (features_test):", features_test.shape)

print(features_temp.head())
features_train.head()

Tamaño del conjunto de entrenamiento (features_train): (2249, 4)
Tamaño del conjunto de validación (features_valid): (482, 4)
Tamaño del conjunto de prueba (features_test): (483, 4)
      calls  minutes  messages   mb_used
1727  104.0   739.28      55.0  21940.11
2885   67.0   469.58      63.0   7952.07
2730   74.0   549.14      20.0  17160.61
899    31.0   205.14      22.0  10250.58
2119   26.0   172.87      16.0   6431.26


calls  minutes  messages   mb_used
1375   67.0   429.17      11.0  13946.66
3155   64.0   431.79      14.0  14675.42
2345  111.0   721.82       0.0  13559.68
2637   42.0   274.71      13.0  20319.36
2272   93.0   658.89      38.0  17655.47

In [6]:
# Importar las librerías necesarias para los modelos y la evaluación
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Modelo Random Forest (modelo que funciona bien tanto con datos estructurados como no estructurados y puede capturar patrones complejos)
rf_model = RandomForestClassifier(random_state=54321)  # Establece random_state para garantizar reproducibilidad
rf_model.fit(features_train, target_train)  # Entrena el modelo con el conjunto de datos de entrenamiento
rf_predictions = rf_model.predict(features_valid)  # Realiza predicciones en el conjunto de validación
rf_accuracy = accuracy_score(target_valid, rf_predictions)  # Evalúa la precisión del modelo

# Modelo Decision Tree (es un modelo más simple que también puede manejar ambos tipos de datos pero puede sobreajustarse si no se controla)
dt_model = DecisionTreeClassifier(random_state=54321)  # Establece random_state para reproducibilidad
dt_model.fit(features_train, target_train)  # Entrena el modelo con el conjunto de entrenamiento
dt_predictions = dt_model.predict(features_valid)  # Realiza predicciones en el conjunto de validación
dt_accuracy = accuracy_score(target_valid, dt_predictions)  # Evalúa la precisión del modelo

# Modelo Logistic Regression (un modelo lineal que es más fácil de interpretar pero puede no capturar patrones complejos tan bien como otros)
lr_model = LogisticRegression(random_state=54321, solver='liblinear')  # Establece random_state y el solver para manejar datasets pequeños
lr_model.fit(features_train, target_train)  # Entrena el modelo con el conjunto de datos de entrenamiento
lr_predictions = lr_model.predict(features_valid)  # Realiza predicciones en el conjunto de validación
lr_accuracy = accuracy_score(target_valid, lr_predictions)  # Evalúa la precisión del modelo

# Imprimir los resultados de precisión de cada modelo en el conjunto de validación
rf_accuracy, dt_accuracy, lr_accuracy


(0.8091286307053942, 0.7510373443983402, 0.6721991701244814)

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Bien hecho! Solo algunos puntos que debes considerar en esta parte:
    
- Procura importar las librerías al inicio del notebook
- Cuando imprimas algún dato es mejor que uses strings formateados y que agregues alguna frase que de a entender el valor que estás mostrando ya que solo mostrar números puede ser confuso
</div>

Hasta el momento los resultados de la exactitud de cada modelo usado (sin modificar los hiperparametros) son los siguientes:

Random Forest: 80.91%
Decision Tree: 75.10%
Logistic Regression: 67.22%

El modelo de Random Forest ha obtenido el mejor resultado hasta ahora.

In [7]:
# Ahora estaremos iterando con los hyperparametros para ver si los modelos descritos mejoran.
#Inicializamos el mejor modelo encontrado con los hiperparámetros óptimos de RANDOM FOREST,seleccione este modelo ya que tuvo la mejor exactitud
#en los codigos anteriores.

# Importamos las librerías necesarias
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Lista de hiperparámetros para probar
n_estimators_list = [50, 100, 200]
max_depth_list = [5, 10, 15, 20]
min_samples_split_list = [2, 5, 10]
min_samples_leaf_list = [1, 2, 4]

# Variables para almacenar los mejores resultados
best_accuracy = 0
best_params = {}

# Iteramos sobre todas las combinaciones de hiperparámetros
for n_estimators in n_estimators_list:
    for max_depth in max_depth_list:
        for min_samples_split in min_samples_split_list:
            for min_samples_leaf in min_samples_leaf_list:
                # Inicializamos el modelo Random Forest con los hiperparámetros actuales
                rf_model = RandomForestClassifier(
                    random_state=54321, 
                    n_estimators=n_estimators, 
                    max_depth=max_depth, 
                    min_samples_split=min_samples_split, 
                    min_samples_leaf=min_samples_leaf
                )
                
                # Entrenamos el modelo
                rf_model.fit(features_train, target_train)
                
                # Hacemos predicciones en el conjunto de validación
                valid_predictions = rf_model.predict(features_valid)
                
                # Calculamos la precisión en el conjunto de validación
                accuracy = accuracy_score(target_valid, valid_predictions)
                
                # Guardamos los mejores hiperparámetros si encontramos una mejor precisión
                if accuracy > best_accuracy:
                    best_accuracy = accuracy
                    best_params = {
                        'n_estimators': n_estimators,
                        'max_depth': max_depth,
                        'min_samples_split': min_samples_split,
                        'min_samples_leaf': min_samples_leaf
                    }

# Imprimimos los mejores hiperparámetros y la precisión obtenida
print("Mejores hiperparámetros:", best_params)
print("Precisión del modelo ajustado en el conjunto de validación:", best_accuracy)


Mejores hiperparámetros: {'n_estimators': 50, 'max_depth': 10, 'min_samples_split': 2, 'min_samples_leaf': 1}
Precisión del modelo ajustado en el conjunto de validación: 0.8112033195020747


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo probando diferentes hiperparámetros en tu modelo!
</div>

In [8]:
from sklearn.model_selection import cross_val_score

#Inicializamos el mejor modelo encontrado con los hiperparámetros óptimos de RANDOM FOREST,seleccione este modelo ya que tuvo la mejor exactitud
#en los codigos anteriores.
best_rf_model = RandomForestClassifier(
    random_state=54321,
    n_estimators=best_params['n_estimators'],
    max_depth=best_params['max_depth'],
    min_samples_split=best_params['min_samples_split'],
    min_samples_leaf=best_params['min_samples_leaf']
)

#validación cruzada con 5 particiones
cv_scores = cross_val_score(best_rf_model, features_train, target_train, cv=5, scoring='accuracy')

#resultados de la validación cruzada
print("Puntajes de validación cruzada en cada partición:", cv_scores)
print("Precisión media de la validación cruzada:", cv_scores.mean())

Puntajes de validación cruzada en cada partición: [0.81777778 0.81777778 0.79555556 0.82666667 0.79955457]
Precisión media de la validación cruzada: 0.8114664686958672


La validación cruzada (cross-validation) nos permite evaluar el modelo de manera más robusta al dividir el conjunto de entrenamiento en múltiples subconjuntos y evaluar el rendimiento del modelo en cada uno de ellos. De esta manera, podemos asegurarnos de que el modelo no dependa excesivamente de una sola partición de los datos.



<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Muy bien aplicada la validación cruzada, obtuviste una métrica por encima del umbral propuesto
    
</div>

# CONCLUSION:
 
###Random Forest parece ser el modelo más adecuado para esta tarea, logrando una alta precisión y adaptándose bien a los patrones en los datos.
###Decision Tree ha mostrado un buen rendimiento, pero tiene más riesgo de sobreajuste debido a su capacidad para crear reglas muy específicas.
###Regresión Logística, aunque es un modelo simple y fácil de interpretar, no parece ser lo suficientemente flexible para capturar la complejidad de los datos.

Ajusté los hiperparámetros de Decision Tree y Regresión Logística para buscar mejoras en la precisión.
Apliqué validación cruzada para asegurarme de que los resultados no dependan únicamente de la partición de validación y que el modelo generalice bien a nuevos datos.
Finalmente, comparé los modelos ajustados en el conjunto de prueba para hacer la selección final del modelo más robusto.


<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con tu proyecto Eduardo! Entrenaste los modelos correctamente y los fuiste mejorando cambiando los hiperparámetros con varias técnicas logrando así una métrica por encima del umbral que se propuso para este proyecto.
    
Sin embargo para tus próximos proyectos te invito a considerar los siguientes puntos:
    
- Realizar un EDA de los datos con los que vas a entrenar tu modelo, puedes generar gráficas y explorar las diferentes variables ya que así entenderás mejor los datos y sabrás que es lo que puede afectar tu modelo
- Procura utilizar más las características de Markdown para incluir títulos o comentarios en cada sección de tu proyecto
- Procura que la importación de librerías que vas a utilizar sean todas al inicio del notebook para mantener un mejor orden
    
Saludos!
</div>